"role attribute has valid value" (674b10): Prepare for TF review (#1350)
act-rules/act-rules.github.io: _rules/role-attribute-valid-value-674b10.md

## Applicability

> Any [`role` attribute][role attribute] for which all the following are true:

Reviewer: Linking attributes is only necessary if they are ambiguous. There is only one type of role attribute in HTML.
```suggestion
Any `role` attribute for which all the following are true:
```

- the attribute has a value that is neither empty ("") nor only [ASCII whitespace][]; and
- the attribute is specified on an HTML or SVG element that is [included in the accessibility tree][].

## Expectation

Each test target has at least one token which is a valid value corresponding to a non-abstract role from [WAI-ARIA Specifications][].

## Assumptions

> The ARIA `role` is being used to comply to WCAG. If compliance to WCAG is done in other ways, this rule may fail while [Success Criterion 4.1.2 Name, Role, Value][sc412] is still satisfied.

Reviewer: I'm not sure this made the assumption any clearer.... Perhaps we can do:

```suggestion
If the `role` attribute is used for something other than giving elements an [explicit role][], the success criterion could be satisfied even if the rule failed. For example, if for some reason the `role` attribute contains a date, this rule will fail. The implicit role of such an element may mean be sufficient to satisfy the success criterion.```

#### Passed Example 1

> This [`role` attribute][role attribute] contains one token, and this token is a valid [WAI-ARIA role][].

Reviewer: ```suggestion
This `role` attribute contains one token, and this token is a valid [WAI-ARIA role][].
```

```html
<input type="text" role="textbox" />
```

#### Passed Example 2

> This [`role` attribute][role attribute] contains two tokens, and these tokens are both valid [WAI-ARIA roles][wai-aria role].

Reviewer: ```suggestion
This `role` attribute contains two tokens, and these tokens are both valid [WAI-ARIA roles][wai-aria role].
```

```html
<span role="button link"></span>
```

#### Passed Example 3

> This [`role` attribute][role attribute] contains two tokens, and one of these tokens is a valid [WAI-ARIA role][].

Reviewer: ```suggestion
This `role` attribute contains two tokens, and one of these tokens is a valid [WAI-ARIA role][].
```

```html

> <img role="img xyz" src="/test-assets/shared/w3c-logo.png" alt="W3C logo" />

Author: This example could be upgraded a bit by using another value than "xyz" that looks like a "plausible typo".
I preferred not touching the examples to keep implementation data.

Reviewer: I like that idea. I think we should just make that change.

#### Failed Example 1

> This [`role` attribute][role attribute] contains one token, but this token is not a valid role in any of the [WAI-ARIA specifications][].

Reviewer: ```suggestion
This `role` attribute contains one token, but this token is not a valid role in any of the [WAI-ARIA specifications][].
```

```html

> <input role="invalid" value="123" />

Author: Same. A more plausible failing value could be used.

```

#### Failed Example 2

> This [`role` attribute][role attribute] contains two tokens, but none of these tokens is a valid role in any of the [WAI-ARIA specifications][].

Reviewer: ```suggestion
This `role` attribute contains two tokens, but none of these tokens is a valid role in any of the [WAI-ARIA specifications][].
```

```html
<input type="text" role="invalid role" />
```

#### Failed Example 3

> This [`role` attribute][role attribute] contains one token, but this token is not a valid role in any of the [WAI-ARIA specifications][].

Reviewer: ```suggestion
This `role` attribute contains one token, but this token is not a valid role in any of the [WAI-ARIA specifications][].
```

```html

> <input type="text" role="#" />

Author: I am not sure this example is really needed as it is essentially the same problem as Failed Example 1.
Should we remove it?

Reviewer: Agreed, lets take it out.

> 

Author: Inapplicable examples have been moved around to mimic the order of conditions in the applicability.
They haven't change, despite what Github believes 😉

#### Inapplicable Example 1

> There is no [`role` attribute][role attribute].

Reviewer: ```suggestion
There is no `role` attribute.
```

```html
<div>Some Content</div>
```

#### Inapplicable Example 2

> This [`role` attribute][role attribute] has no value.

Reviewer: ```suggestion
This `role` attribute has no value.
```

```html
<div role>Some Content</div>
```

#### Inapplicable Example 3

> This [`role` attribute][role attribute] is empty ("").

Reviewer: ```suggestion
This `role` attribute is empty ("").
```

```html
<div role="">Some Content</div>
```

#### Inapplicable Example 4

> This [`role` attribute][role attribute] is only [ASCII whitespace][].

Reviewer: ```suggestion
This `role` attribute is only [ASCII whitespace][].
```

```html
<input type="text" role=" " />
```

#### Inapplicable Example 5

> This [`role` attribute][role attribute] is specified on an element which is not [included in the accessibility tree][].

Reviewer: ```suggestion
This `role` attribute is specified on an element which is not [included in the accessibility tree][].
```

```html
<div aria-hidden="true" role="banner">Some Content</div>
```

[ascii whitespace]: https://infra.spec.whatwg.org/#ascii-whitespace 'Definition of ASCII whitespace'
[included in the accessibility tree]: #included-in-the-accessibility-tree 'Definition of included in the accessibility tree'

> [role attribute]: https://www.w3.org/TR/role-attribute/ 'Specification of the Role attribute'

Reviewer: ```suggestion
```
